Blog post for OpenTelemetry Generative AI updates (#5575)
open-telemetry/opentelemetry.io: content/en/blog/2024/otel-generative-ai/index.md

---
title: OpenTelemetry for Generative AI
linkTitle: OpenTelemetry for Generative AI
date: 2024-11-09
author: >-
  [Drew Robbins](https://github.com/drewby) (Microsoft),  [Liudmila
  Molkova](https://github.com/lmolkova) (Microsoft)
---

As organizations increasingly adopt Large Language Models (LLMs) and other
generative AI technologies, ensuring reliable performance, efficiency, and

> safety is essential. Enhanced observability—tracking AI operations, behaviors,

Reviewer: > ensuring reliable performance, efficiency, and safety is essential.

Why?

Author: I updated the intro to answer Why.

operational details. For generative AI, these conventions streamline monitoring,
troubleshooting, and optimizing AI models by standardizing attributes such as
model parameters, response metadata, and token usage. This consistency supports
better observability across tools, environments, and APIs, helping organizations
track performance, cost, and safety with ease.

The **Instrumentation Library** is being developed within the OpenTelemetry
Python Contrib project to automate telemetry collection for generative AI
applications. The first release is a Python library, given Python’s widespread
use in AI development. Designed to integrate seamlessly with OpenAI’s API, this
library captures spans and events, gathering essential data like model
inputs, response metadata, and token usage in a structured format.

## Key Signals for Generative AI

The
[Semantic Conventions for Generative AI](https://github.com/open-telemetry/semantic-conventions/tree/v1.28.0/docs/gen-ai)
focus on capturing insights into AI model behavior through three primary
signals: Spans, Metrics, and Events.

**Spans: Tracing Model Interactions**  

> Spans track each model interaction’s lifecycle, covering input parameters (e.g.,
> temperature, top_p) and response details like token count or errors. They

Reviewer: ```suggestion
Spans track each model interaction’s lifecycle, covering input parameters (for
example, temperature, top_p) and response details like token count or errors. They
```

provide visibility into each request, aiding in identifying bottlenecks and
analyzing the impact of settings on model output.

**Metrics: Monitoring Usage and Performance**  
Metrics aggregate high-level indicators like request volume, latency, and
token counts, essential for managing costs and performance. This data is
particularly critical for API-dependent AI applications with rate limits and
cost considerations.

**Events: Capturing Detailed Interactions**  
Events log detailed moments during model execution, such as user prompts and
model responses, providing a granular view of model interactions. These insights
are invaluable for debugging and optimizing AI applications where unexpected
behaviors may arise.

Together, these signals provide a comprehensive monitoring framework, enabling
better cost management, performance tuning, and request tracing.

> **Extending Observability with Vendor-Specific Attributes** The Semantic

Reviewer: Is the part in bold meant to be a title? Consider using ### markers for each of the subsections.

---
title: OpenTelemetry for Generative AI
linkTitle: OpenTelemetry for Generative AI

> date: 2024-11-09

Reviewer 2: Putting this comment here to keep an eye on setting the date right when we finally publish. Do not resolve.

```suggestion
date: 2024-11-09
```

Reviewer 2: I would like to have @chalin re-review the PR, and we shouldn't publish this week anyways due to thanksgiving

author: >-
  [Drew Robbins](https://github.com/drewby) (Microsoft),  [Liudmila
  Molkova](https://github.com/lmolkova) (Microsoft)

> ---

Reviewer 2: ```suggestion
issue: 
sig: SIG GenAI Observability
---
```

@drewby can you create an issue with the same comment as you have here in the PR description, so we can link it to this blog post. Since this PR has everything needed (SIG + sponsors) we can keep it, but we want to establish this practice so let's set a good example

Author: will do

As organizations increasingly adopt Large Language Models (LLMs) and other
generative AI technologies, ensuring reliable performance, efficiency, and
safety is essential to meet user expectations, optimize resource costs, and
safeguard against unintended outputs. Effective observability for AI operations,
behaviors, and outcomes can help meet these goals. OpenTelemetry is being
enhanced to support these needs specifically for generative AI.

Two primary assets are in development to make this possible: **Semantic

> Conventions** and an **Instrumentation Library**.

Reviewer 2: ```suggestion
Conventions** and an **Instrumentation Libraries**.
```

Is there only one library that is developed or are there multiple libraries?

Author: There is only one at this moment. We focused on Python first.

Reviewer 2: then maybe write this here already:

```suggestion
Conventions** and an **Instrumentation Library** for OpenAI's APIs:
```

Reviewer 3: If we made this change, it would conflate the GenAI working group assets with OpenAI, which I think doesn't follow later. Also, it feels less attractive to folks who are not OpenAI based, yet should be welcome into the team. Also, the specs are used for non-openai right now. It is just that the first library instrumentation in the shared codebase is openai.

If we want to hint what's first, I think we should make that a second sentence. wdyt?

```suggestion
Conventions** and an **Instrumentation Library**. The first instrumentation library targets OpenAI in Python.
```

> **Semantic Conventions** establish standardized guidelines for how telemetry

Reviewer 2: ```suggestion
[**Semantic Conventions** ](https://opentelemetry.io/docs/concepts/semantic-conventions/) establish standardized guidelines for how telemetry
```

I'm fine with the inline explanation (and maybe we can even take some if and improve the existing concept page), but link to docs pages where possible, so people can dig deeper if they want to learn more

data is structured and collected across platforms, defining inputs, outputs, and
operational details. For generative AI, these conventions streamline monitoring,
troubleshooting, and optimizing AI models by standardizing attributes such as
model parameters, response metadata, and token usage. This consistency supports
better observability across tools, environments, and APIs, helping organizations
track performance, cost, and safety with ease.

> The **Instrumentation Library** is being developed within the OpenTelemetry
> Python Contrib project to automate telemetry collection for generative AI
> applications. The first release is a Python library, given Python’s widespread

Reviewer 2: ```suggestion
The [**Instrumentation Library**](https://opentelemetry.io/docs/specs/otel/overview/#instrumentation-libraries) is being developed within the [OpenTelemetry Python Contrib](https://github.com/open-telemetry/opentelemetry-python-contrib) project to automate telemetry collection for generative AI
applications. The first release is a Python library, given Python’s widespread
```

> use in AI development. Designed to integrate seamlessly with OpenAI’s API, this

Reviewer 2: Same question as above, is there only one library (for OpenAI) or multiple? If there is only one make it much clearer that "the instrumentation library" is "the instrumentation library for OpenAI"

library captures spans and events, gathering essential data like model
inputs, response metadata, and token usage in a structured format.

## Key Signals for Generative AI

The

> [Semantic Conventions for Generative AI](https://github.com/open-telemetry/semantic-conventions/tree/v1.28.0/docs/gen-ai)

Reviewer 2: ```suggestion
[Semantic Conventions for Generative AI](/docs/specs/semconv/gen-ai/)
```

Author: We were thinking to pin the version here to prevent potential broken links in the future. cc @lmolkova

Reviewer 3: agree on pinning as this has drifted in the past, unless there's a process in the website to find bad links and automatically raise PRs to retarget them.

Reviewer 2: We have a process to find bad links automatically, also this is a blog post not a documentation page, so it's expected for them to decay, see https://github.com/open-telemetry/opentelemetry.io/issues/3493, cc @chalin

focus on capturing insights into AI model behavior through three primary

> signals: Spans, Metrics, and Events.

Reviewer 2: ```suggestion
signals: [Traces](/docs/concepts/signals/traces/), [Metrics](/docs/concepts/signals/metrics/), and [Logs](/docs/concepts/signals/logs/).
```

Re events: I am fine with that as well, but people will stumble upon that and ask themselves "why events and not logs", so you would need to add some explanation and links about it. I know that this is a highly debated subject within the community but this blog post will be consumed by a lot of people that do not know all the things happening within OTel so we need to be cautious to put words out there without adding some more context.

> ### Spans: Tracing Model Interactions

Reviewer 2: ```suggestion
### Traces: Tracing Model Interactions
```

> Spans track each model interaction’s lifecycle, covering input parameters (for
> example, temperature, top_p) and response details like token count or errors.

Reviewer 2: ```suggestion
Traces track each model interaction’s lifecycle, covering input parameters (for
example, temperature, top_p) and response details like token count or errors.
```

They provide visibility into each request, aiding in identifying bottlenecks and
analyzing the impact of settings on model output.

> cd 

Reviewer 2: ```suggestion

```

### Metrics: Monitoring Usage and Performance

Metrics aggregate high-level indicators like request volume, latency, and
token counts, essential for managing costs and performance. This data is
particularly critical for API-dependent AI applications with rate limits and
cost considerations.

### Events: Capturing Detailed Interactions

Events log detailed moments during model execution, such as user prompts and
model responses, providing a granular view of model interactions. These insights
are invaluable for debugging and optimizing AI applications where unexpected

> behaviors may arise.

Reviewer 2: See my comment above around logs vs events

Together, these signals provide a comprehensive monitoring framework, enabling

> better cost management, performance tuning, and request tracing.

Reviewer 2: Since you used sub headings this feels like a part of "Events: Capturing Detailed Interactions", you can move it up or think about explaining those three in bullet points instead.

## Building the Python Instrumentation Library for OpenAI

This Python-based library for OpenTelemetry captures key telemetry signals for
OpenAI models, providing developers with an out-of-the-box observability
solution tailored to AI workloads. The library,
[hosted within the OpenTelemetry Python Contrib repository](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/opentelemetry-instrumentation-openai-v2%3D%3D2.0b0/instrumentation-genai/opentelemetry-instrumentation-openai-v2),
automatically collects telemetry from OpenAI model interactions, including
request and response metadata and token usage.

As generative AI applications grow, additional instrumentation libraries for
other languages will follow, extending OpenTelemetry support across more tools
and environments. The current library’s focus on OpenAI highlights its
popularity and demand within AI development, making it a valuable initial
implementation.

### Example Usage

Here’s an example of using the OpenTelemetry Python library to monitor a
generative AI application with the OpenAI client. Make sure you first install
the library:

```bash
pip install opentelemetry-instrumentation-openai-v2
```

> Then include the following code in your Python application.

Reviewer 2: ```suggestion
Then include the following code in your Python application:
```

```python
from openai import OpenAI
from opentelemetry.instrumentation.openai_v2 import OpenAIInstrumentor

OpenAIInstrumentor().instrument()

client = OpenAI()
response = client.chat.completions.create(
    model="gpt-4-mini",
    messages=[{"role": "user", "content": "Write a short poem on OpenTelemetry."}],
)

# The library captures telemetry, including request and response metadata, token usage, and more.
```

With this simple instrumentation, one can begin capture traces from their
generative AI application. Here is an example from the
[Aspire Dashboard](https://learn.microsoft.com/dotnet/aspire/fundamentals/dashboard/standalone?tabs=bash)
for local debugging.

![Chat trace in Aspire Dashboard](aspire-dashboard-trace.png)

Here is a similar trace captured in
[Jaeger](https://www.jaegertracing.io/docs/next-release-v2/getting-started/#running):

> ![Chat trace in Jaeger](jaeger-trace.png)

Reviewer 2: Limit this to the Jaeger example (see my comment above around prefering CNCF projects over other projects).

Author: Jeager doesn't show the Events/Logs we are using for Content Capture. Unless I'm missing how to view these events in Jaeger. @lmolkova, do you know a way?

---
title: OpenTelemetry for Generative AI
linkTitle: OpenTelemetry for Generative AI
date: 2024-11-09
author: >-
  [Drew Robbins](https://github.com/drewby) (Microsoft),  [Liudmila
  Molkova](https://github.com/lmolkova) (Microsoft)
issue: [#5581](https://github.com/open-telemetry/opentelemetry.io/issues/5581)

> sig: SIG GenAI Observability
> ---

Reviewer 2: ```suggestion
sig: SIG GenAI Observability
cSpell:ignore: genai liudmila molkova
---
```

safety is essential to meet user expectations, optimize resource costs, and
safeguard against unintended outputs. Effective observability for AI operations,
behaviors, and outcomes can help meet these goals. OpenTelemetry is being
enhanced to support these needs specifically for generative AI.

Two primary assets are in development to make this possible: **Semantic
Conventions** and an **Instrumentation Library**.

[**Semantic Conventions**](https://opentelemetry.io/docs/concepts/semantic-conventions/)
establish standardized guidelines for how telemetry data is structured and
collected across platforms, defining inputs, outputs, and operational details.
For generative AI, these conventions streamline monitoring, troubleshooting, and
optimizing AI models by standardizing attributes such as model parameters,
response metadata, and token usage. This consistency supports better
observability across tools, environments, and APIs, helping organizations track
performance, cost, and safety with ease.

The
[**Instrumentation Library**](https://opentelemetry.io/docs/specs/otel/overview/#instrumentation-libraries)
is being developed within the

> [OpenTelemetry Python Contrib](https://github.com/open-telemetry/opentelemetry-python-contrib)

Reviewer 4: how about link to https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation-genai?

Author: We have the link later in the doc when we talk about the specifics of the library.

Reviewer 4: Yes, Line 94 is a link for a specified link for openai instrumentation, but here should be the folder link of `instrumentation-genai`?

```suggestion
[OpenTelemetry Python Contrib](https://github.com/open-telemetry/opentelemetry-python-contrib) under `[instrumentation-genai](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation-genai)`
```

[Events](https://opentelemetry.io/docs/specs/otel/logs/event-api/).

Together, these signals provide a comprehensive monitoring framework, enabling
better cost management, performance tuning, and request tracing.

### Traces: Tracing Model Interactions

Traces track each model interaction’s lifecycle, covering input parameters (for
example, temperature, top_p) and response details like token count or errors.
They provide visibility into each request, aiding in identifying bottlenecks and
analyzing the impact of settings on model output.

### Metrics: Monitoring Usage and Performance

Metrics aggregate high-level indicators like request volume, latency, and token
counts, essential for managing costs and performance. This data is particularly
critical for API-dependent AI applications with rate limits and cost
considerations.

### Events: Capturing Detailed Interactions

Events log detailed moments during model execution, such as user prompts and
model responses, providing a granular view of model interactions. These insights
are invaluable for debugging and optimizing AI applications where unexpected
behaviors may arise.

> > Note that we decided to use the newer Events API

Author: @svrnm I've added this blockquote as a note on our use of the Events API. Does this resolve the issue you are raising around logs vs events?

Reviewer 2: Yes, that's perfect 🙌

> > Note that we decided to use the newer Events API
> > (https://opentelemetry.io/docs/specs/otel/logs/event-api/) specification in
> > the Semantic Conventions for Generative AI. The events API allows for us to
> > define specific
> > [semantic conventions](https://opentelemetry.io/docs/specs/semconv/general/events/)
> > for the user prompts and model responses that we capture.

Reviewer 2: ```suggestion
{{% alert title="Note" color="info" %}}

Note that we decided to use the newer Events API
(https://opentelemetry.io/docs/specs/otel/logs/event-api/) specification in
the Semantic Conventions for Generative AI. The events API allows for us to
define specific
[semantic conventions](https://opentelemetry.io/docs/specs/semconv/general/events/)
for the user prompts and model responses that we capture.

{{% /alert %}}
```

While the `>` accomplish something similar, let's use the [alert shortcode](https://www.docsy.dev/docs/adding-content/shortcodes/#alert)

Reviewer 3: I feel strongly we should say "Log events API" as it is a subclass of logging, and if you don't configure logging properly nothing will work. This is particularly important because status quo before was span events.

Author: @codefromthecrypt I hear you, however I don't see it referred to that way elsewhere? Maybe it is, I just didn't see it.

Reviewer 2: The [Events API](https://opentelemetry.io/docs/specs/otel/logs/event-api/) states that _"this existing Events API and once it is defined and implemented, the Events API usage will be migrated, deprecated, renamed and eventually removed."_ 

Reading this note (we use the new events api) and then reading that it is going to be deprecated, is confusing for an end user. 

@lmolkova @tedsuo any guidance here?

Reviewer 3: @drewby it is in the log directory and many places it has the word log in the doc. We should fix the heading in the doc in my opinion if that's the only way to sort out this blog to highlight that this is log events not span events

Reviewer 3: signal wise, we have log, metrics, spans, profile. If we want instead to literally say "events api" we should introduce it as a log signal. I personally jumping straight to "events api" doesn't prepare folks that this uses log signals, and a single word helps untangle this. EOF!

Reviewer 5: Event API is being merged into Logs API - https://github.com/open-telemetry/opentelemetry-specification/blob/v1.39.0/specification/logs/api.md#emit-an-event.

So a future-proof way would be to say 

> we decided to use the log-based events

or

> we decided to use events emitted with Logs API

Author: I've updated the Alert text based on @lmolkova feedback.

like OpenAI and Azure Inference API, ensuring telemetry captures both general
and provider-specific details. This added flexibility supports multi-platform
monitoring and in-depth insights.

## Building the Python Instrumentation Library for OpenAI

This Python-based library for OpenTelemetry captures key telemetry signals for
OpenAI models, providing developers with an out-of-the-box observability
solution tailored to AI workloads. The library,
[hosted within the OpenTelemetry Python Contrib repository](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/opentelemetry-instrumentation-openai-v2%3D%3D2.0b0/instrumentation-genai/opentelemetry-instrumentation-openai-v2),
automatically collects telemetry from OpenAI model interactions, including
request and response metadata and token usage.

As generative AI applications grow, additional instrumentation libraries for
other languages will follow, extending OpenTelemetry support across more tools
and environments. The current library’s focus on OpenAI highlights its
popularity and demand within AI development, making it a valuable initial
implementation.

> ### Example Usage

Reviewer 6: Would be good to show the metrics being captures as well. The name of the metric and the attributes captured along with it.

Author: We are not capturing metrics in the instrumentation library yet. Good place to contribute! ;)

Here’s an example of using the OpenTelemetry Python library to monitor a
generative AI application with the OpenAI client. Make sure you first install
the library:

```bash
pip install opentelemetry-instrumentation-openai-v2
```

> Then include the following code in your Python application:

Reviewer 3: this example is currently misleading, as it looks like it is zero code. The large setup logic is not visible in the blog. This is why I strongly don't agree with the optimization of time vs clean.

There are new facets people need to plumb for this to work as expected. People in genai expect to see prompt/completion data. We should make it work (zero code approach) or make very visible the code required to have that happen.

for local debugging.

![Chat trace in Aspire Dashboard](aspire-dashboard-trace.png)

Here is a similar trace captured in
[Jaeger](https://www.jaegertracing.io/docs/next-release-v2/getting-started/#running):

![Chat trace in Jaeger](jaeger-trace.png)

It's also easy to capture the content history of the chat for debugging and
improving your application. Simply set the environment variable
`OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT` as follows:

```bash
export OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT=True
```

This will turn on content capture which collects OpenTelemetry events containing
the payload:

![Content Capture Aspire Dashboard](aspire-dashboard-content-capture.png)

## Join Us in Shaping the Future of Generative AI Observability

Community collaboration is key to OpenTelemetry’s success. We invite developers,
AI practitioners, and organizations to contribute, share feedback, or
participate in discussions. Explore the OpenTelemetry Python Contrib project,
contribute code, or help shape observability for AI as it continues to evolve.
More information can be found at the

> [Generative AI Observability project page](https://github.com/open-telemetry/community/blob/main/projects/gen-ai.md).

Reviewer 4: @drewby How about highlighting the current contributors at the end of the blog? This way, readers can quickly see who’s actively involved in this SIG.

```suggestion
[Generative AI Observability project page](https://github.com/open-telemetry/community/blob/main/projects/gen-ai.md), we now have contributors from OpenLIT, Langtrace, Elastic, MicroSoft, Traceloop, IBM etc., welcome to join the community!
```

Reviewer 7: Happy to include Scorecard as a supporter and adopter of GenAI OTEL!

Author: > @drewby How about highlighting the current contributors at the end of the blog? This way, readers can quickly see who’s actively involved in this SIG.

It's a good suggestion and I'd be delighted to include how we've had great collaboration and contribution across all these companies. 

@svrnm I'm not sure how this fits with the community/blog guidelines?

Reviewer 2: I think that's fine, but I would suggest to order them alphabetically

Reviewer 4: Awesome, thanks @drewby !

from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.grpc._log_exporter import OTLPLogExporter

trace.set_tracer_provider(TracerProvider())
trace.get_tracer_provider().add_span_processor(
    BatchSpanProcessor(OTLPSpanExporter())
)

_logs.set_logger_provider(LoggerProvider())
_logs.get_logger_provider().add_log_record_processor(BatchLogRecordProcessor(OTLPLogExporter()))
_events.set_event_logger_provider(EventLoggerProvider())

from openai import OpenAI
from opentelemetry.instrumentation.openai_v2 import OpenAIInstrumentor

OpenAIInstrumentor().instrument()

client = OpenAI()
response = client.chat.completions.create(
    model="gpt-4-mini",
    messages=[{"role": "user", "content": "Write a short poem on OpenTelemetry."}],
)

# The library captures telemetry, including request and response metadata, token usage, and more.
```

With this simple instrumentation, one can begin capture traces from their
generative AI application. Here is an example from the
[Aspire Dashboard](https://learn.microsoft.com/dotnet/aspire/fundamentals/dashboard/standalone?tabs=bash)
for local debugging.

> ![Chat trace in Aspire Dashboard](aspire-dashboard-trace.png)

Reviewer 3: I noticed both of these traces are not from the sample code, as they have spans that it wouldn't produce. Is this ok?

implementation.

### Example Usage

Here’s an example of using the OpenTelemetry Python library to monitor a
generative AI application with the OpenAI client. Make sure you first install
the library:

```bash
pip install opentelemetry-instrumentation-openai-v2
```

Set the following environment variables, updating the endpoint and protocol as
appropriate:

```bash
OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
OTEL_SERVICE_NAME=python-opentelemetry-openai

# Set to false or remove to disable log events
OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT=true
OTEL_PYTHON_LOGGING_AUTO_INSTRUMENTATION_ENABLED=true
OTEL_LOGS_EXPORTER=otlp_proto_http
```

Then include the following code in your Python application:

```python
# NOTE: OpenTelemetry Python Log Events APIs is in beta
from opentelemetry import _events
from opentelemetry.sdk._events import EventLoggerProvider
_events.set_event_logger_provider(EventLoggerProvider())

from openai import OpenAI
from opentelemetry.instrumentation.openai_v2 import OpenAIInstrumentor

OpenAIInstrumentor().instrument()

client = OpenAI()
response = client.chat.completions.create(

>     model="gpt-4-mini",

Reviewer 3: At least for me, I had to switch this for my account to work

```suggestion
    model="gpt-4o-mini",
```

    messages=[{"role": "user", "content": "Write a short poem on OpenTelemetry."}],
)

# The library captures telemetry, including request and response metadata, token usage, and more.
```

With this simple instrumentation, one can begin capture traces from their
generative AI application. Here is an example from the
[Aspire Dashboard](https://learn.microsoft.com/dotnet/aspire/fundamentals/dashboard/standalone?tabs=bash)
for local debugging.

![Chat trace in Aspire Dashboard](aspire-dashboard-trace.png)

Here is a similar trace captured in

> [Jaeger](https://www.jaegertracing.io/docs/next-release-v2/getting-started/#running):

Reviewer 8: ```suggestion
[Jaeger](https://www.jaegertracing.io/docs/2.0/getting-started/#running):
```

For generative AI, these conventions streamline monitoring, troubleshooting, and
optimizing AI models by standardizing attributes such as model parameters,
response metadata, and token usage. This consistency supports better
observability across tools, environments, and APIs, helping organizations track
performance, cost, and safety with ease.

The
[**Instrumentation Library**](https://opentelemetry.io/docs/specs/otel/overview/#instrumentation-libraries)
is being developed within the
[OpenTelemetry Python Contrib](https://github.com/open-telemetry/opentelemetry-python-contrib)
under
[instrumentation-genai](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation-genai)
project to automate telemetry collection for generative AI applications. The
first release is a Python library for instrumenting OpenAI client calls, given
Python's widespread use in AI development and the popularity of OpenAI. Designed
to integrate seamlessly with OpenAI's API, this library captures spans and
events, gathering essential data like model inputs, response metadata, and token
usage in a structured format.

> 

Reviewer 9: can we say that this article is not covering the model server? And maybe call to action for model server authors to look into standardizing metrics they expose?

implementation.

### Example Usage

Here’s an example of using the OpenTelemetry Python library to monitor a
generative AI application with the OpenAI client. Make sure you first install
the library:

```bash
pip install opentelemetry-instrumentation-openai-v2
```

Set the following environment variables, updating the endpoint and protocol as
appropriate:

```bash
OPENAI_API_KEY=<replace_with_your_openai_api_key>

OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
OTEL_SERVICE_NAME=python-opentelemetry-openai
OTEL_LOGS_EXPORTER=otlp_proto_http
OTEL_PYTHON_LOGGING_AUTO_INSTRUMENTATION_ENABLED=true
# Set to false or remove to disable log events
OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT=true
```

Then include the following code in your Python application:

```python
import os
from openai import OpenAI

client = OpenAI()
chat_completion = client.chat.completions.create(
    model=os.getenv("CHAT_MODEL", "gpt-4o-mini"),
    messages=[
        {
            "role": "user",
            "content": "Write a short poem on OpenTelemetry.",
        },
    ],
)
print(chat_completion.choices[0].message.content)

> ```

Reviewer 3: maybe you need a "finally, run the example with `opentelemetry-instrument`" part, because the zero code requires prefixing python with that

implementation.

### Example Usage

Here’s an example of using the OpenTelemetry Python library to monitor a
generative AI application with the OpenAI client. Make sure you first install
the library:

```bash
pip install opentelemetry-instrumentation-openai-v2
```

Set the following environment variables, updating the endpoint and protocol as
appropriate:

```bash
OPENAI_API_KEY=<replace_with_your_openai_api_key>

OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
OTEL_SERVICE_NAME=python-opentelemetry-openai
OTEL_LOGS_EXPORTER=otlp_proto_http
OTEL_PYTHON_LOGGING_AUTO_INSTRUMENTATION_ENABLED=true
# Set to false or remove to disable log events
OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT=true
```

Then include the following code in your Python application:

```python
import os
from openai import OpenAI

client = OpenAI()
chat_completion = client.chat.completions.create(
    model=os.getenv("CHAT_MODEL", "gpt-4o-mini"),
    messages=[
        {
            "role": "user",
            "content": "Write a short poem on OpenTelemetry.",
        },
    ],
)
print(chat_completion.choices[0].message.content)
```

> And then run the example using `opentelemetry-instrument`:

Reviewer 10: I don't believe this example will work that is without installing all of the correct packages. Users would need `opentelemetry-distro` package as well. To make this truly zero code and easier for users to digest, we could add instructions similar to the current [zero-code](https://opentelemetry.io/docs/zero-code/python/#setup) example for Python. Specifically:

```
pip install opentelemetry-distro
opentelemetry-bootstrap -a install
```
Following [this-pr](https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2996), openai instrumentation is automatically added to `opentelemetry-bootstrap`, which is designed to automatically install instrumentation libraries if users have the underlying instrumented library installed (openai in this case). 

tldr: The current example will not actually work properly without `opentelemetry-distro` installed. Replace `pip install opentelemetry-instrumentation-openai-v2` with `pip install opentelemetry-distro` and call `opentelemetry-bootstrap -a install`.

Author: Updated the instructions. Thanks!

---

> title: OpenTelemetry for Generative AI
> linkTitle: OpenTelemetry for Generative AI

Reviewer 11: Drop the stuttered `linkTitle` or shorten it, e.g.:

```suggestion
title: OpenTelemetry for Generative AI
linkTitle: OTel for Generative AI
```

My preference would be to drop it.

date: 2024-11-09
author: >-
  [Drew Robbins](https://github.com/drewby) (Microsoft),  [Liudmila
  Molkova](https://github.com/lmolkova) (Microsoft)

> issue: [#5581](https://github.com/open-telemetry/opentelemetry.io/issues/5581)

Reviewer 11: This is breaking the build I believe. Here's a fix:

```suggestion
issue: https://github.com/open-telemetry/opentelemetry.io/issues/5581
```

Author: thanks!

generative AI application with the OpenAI client.

Install the OpenTelemetry dependencies:

```shell
pip install opentelemetry-distro
opentelemetry-bootstrap -a install
```

Set the following environment variables, updating the endpoint and protocol as
appropriate:

```shell
OPENAI_API_KEY=<replace_with_your_openai_api_key>

OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
OTEL_SERVICE_NAME=python-opentelemetry-openai
OTEL_LOGS_EXPORTER=otlp_proto_http
OTEL_PYTHON_LOGGING_AUTO_INSTRUMENTATION_ENABLED=true
# Set to false or remove to disable log events
OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT=true
```

Then include the following code in your Python application:

```python
import os
from openai import OpenAI

client = OpenAI()
chat_completion = client.chat.completions.create(
    model=os.getenv("CHAT_MODEL", "gpt-4o-mini"),
    messages=[
        {
            "role": "user",
            "content": "Write a short poem on OpenTelemetry.",
        },
    ],
)
print(chat_completion.choices[0].message.content)
```

And then run the example using `opentelemetry-instrument`:

```shell

> opentelemetry-instrument python main.py

Reviewer 10: Is it expected for this to be exporting to console or otlp? Otlp is used by default but if we want a simple get started for customers perhaps this would be better:

```
opentelemetry-instrument --traces_exporter console --metrics_exporter console python .\server_automatic.py
```

Reviewer 10: btw, if user does NOT set up collector and leaves the example as it is to use otlp by default they will get error messages.

Author: I added it as an additional example. We don't have metrics yet, so I just included traces option. Thanks!

Reviewer 10: Nice! If you don't want to include metrics, you have to explicitly set to None.

`opentelemetry-instrument --traces_exporter console --metrics_exporter None python main.py`

Author: In that case, I'll just make it console

OpenAI models, providing developers with an out-of-the-box observability
solution tailored to AI workloads. The library,
[hosted within the OpenTelemetry Python Contrib repository](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/opentelemetry-instrumentation-openai-v2%3D%3D2.0b0/instrumentation-genai/opentelemetry-instrumentation-openai-v2),
automatically collects telemetry from OpenAI model interactions, including
request and response metadata and token usage.

As generative AI applications grow, additional instrumentation libraries for
other languages will follow, extending OpenTelemetry support across more tools
and environments. The current library’s focus on OpenAI highlights its
popularity and demand within AI development, making it a valuable initial
implementation.

### Example Usage

Here’s an example of using the OpenTelemetry Python library to monitor a
generative AI application with the OpenAI client. Make sure you first install
the library:

```bash
pip install opentelemetry-instrumentation-openai-v2
```

Set the following environment variables, updating the endpoint and protocol as
appropriate:

```bash

> OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318

Reviewer 3: TL;DR; IMHO whatever we paste in a bash codeblock, people will expect to work by pasting on their own. This doesn't work, yet even if we fix the pip deps due to not exporting the variables.

maybe worth removing these two `OTEL_EXPORTER` variables and leave it to defaults based on the dependency like `opentelemetry-exporter-otlp-proto-grpc` or `opentelemetry-exporter-otlp-proto-http`. wdyt?

Then, we have only the service name, api key and capture content ones. Also, these need to be exported or set inline. If you don't the python command won't work later. Many people put these into a .env file and use dotenv to load it.

e.g. 

```bash
$ pip install "python-dotenv[cli]"
$ dotenv run -- python main.py
```

However, we can also just inline them like

```bash
$ OTEL_SERVICE_NAME=python-opentelemetry-openai OPENAI_API_KEY=sk-not_tellin OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT=true python main.py
```
```

OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
OTEL_SERVICE_NAME=python-opentelemetry-openai
OPENAI_API_KEY=<replace_with_your_openai_api_key>

# Set to false or remove to disable log events
OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT=true
```

Then include the following code in your Python application:

```python

# NOTE: OpenTelemetry Python Log Events APIs is in beta
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk._logs import LoggerProvider
from opentelemetry.sdk._events import EventLoggerProvider
from opentelemetry.trace import set_tracer_provider
from opentelemetry._logs import set_logger_provider
from opentelemetry._events import set_event_logger_provider
from opentelemetry.sdk.trace.export import SimpleSpanProcessor
from opentelemetry.sdk._logs.export import SimpleLogRecordProcessor

> from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter

Reviewer 3: while grpc is ok I'm not sure if all vendors support that. Also it requires platform deps to build. Should we switch this to http instead?

Reviewer 3: I see now, it is because aspire only supports grpc https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/dashboard/standalone?tabs=bash#configure-opentelemetry-sdk

Reviewer 3: @drewby that link isn't correct afterall, as aspire dashboard also supports http. I suggest changing to http in the final example as that lets you use smaller images like alpine without installing gcc which is needed for the grpc package.

e.g. this works if we use http

```
# Use an alpine image to make the runtime smaller
FROM docker.io/python:3.12.7-alpine3.20
RUN python -m pip install --upgrade pip

COPY /requirements.txt /tmp/requirements.txt
RUN pip install -r /tmp/requirements.txt

COPY main.py /

CMD [ "opentelemetry-instrument", "python", "main.py" ]
```

Reviewer 3: specifically, the following let's you use http on localhost:4318

```bash
docker run --rm -it  -p 18888:18888     -p 4318:18890  --name aspire-dashboard     mcr.microsoft.com/dotnet/aspire-dashboard:9.0
```

```

Then include the following code in your Python application:

```python
import os
from openai import OpenAI

client = OpenAI()
chat_completion = client.chat.completions.create(
    model=os.getenv("CHAT_MODEL", "gpt-4o-mini"),
    messages=[
        {
            "role": "user",
            "content": "Write a short poem on OpenTelemetry.",
        },
    ],
)
print(chat_completion.choices[0].message.content)
```

And then run the example using `opentelemetry-instrument`:

```shell
opentelemetry-instrument python main.py
```

If you do not have a service running to collect telemetry, you can export to the
console using the following:

```shell
opentelemetry-instrument --traces_exporter console --metrics_exporter console python main.py
```

There is a complete example
[available here](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation-genai/opentelemetry-instrumentation-openai-v2/example).

With this simple instrumentation, one can begin capture traces from their
generative AI application. Here is an example from the
[Aspire Dashboard](https://learn.microsoft.com/dotnet/aspire/fundamentals/dashboard/standalone?tabs=bash)

> for local debugging.

Reviewer 5: since we're using `4318` port and `otlp_proto_http` in the example, we should provide the (non-default) way to run aspire dashboard with

```bash
docker run --rm -it -d \
    -p 18888:18888 \
    -p 4317:18889 \
    -p 4318:18890 \
    --name aspire-dashboard \
    mcr.microsoft.com/dotnet/aspire-dashboard:9.0
```

The `-p 4318:18890` is not mentioned in the Aspire docs (as most of otel defaults to grpc protocol on 4317 port).

The alternative is to update the sample here to match python example being updated - 
https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2988/

It should be fine either way, let's just make sure we provide consistent guidance.

Author: I added a docker command for aspire dashboard and jaeger to keep the examples consistent.

Reviewer 2: @lmolkova can you re-review the changes for that?

---
title: OpenTelemetry for Generative AI
linkTitle: OpenTelemetry for Generative AI
date: 2024-11-09
author: >-

>   [Drew Robbins](https://github.com/drewby) (Microsoft),  [Luidmila

Reviewer 5: ```suggestion
  [Drew Robbins](https://github.com/drewby) (Microsoft),  [Liudmila
```

:)

Two primary assets are in development to make this possible: **Semantic Conventions** and an **Instrumentation Library**.

**Semantic Conventions** establish standardized guidelines for how telemetry data is structured and collected across platforms, defining inputs, outputs, and operational details. For generative AI, these conventions streamline monitoring, troubleshooting, and optimizing AI models by standardizing attributes such as model parameters, response metadata, and token usage. This consistency supports better observability across tools, environments, and APIs, helping organizations track performance, cost, and safety with ease.

The **Instrumentation Library** is being developed within the OpenTelemetry Python Contrib project to automate telemetry collection for generative AI applications. The first release is a Python library, given Python’s widespread use in AI development. Designed to integrate seamlessly with OpenAI’s API, this library captures spans, metrics, and events, gathering essential data like model inputs, response metadata, and token usage in a structured format.

## Key Signals for Generative AI

> The [Semantic Conventions for Generative AI](https://github.com/open-telemetry/semantic-conventions/tree/main/docs/gen-ai) focus on capturing insights into AI model behavior through three primary signals: Spans, Metrics, and Events.

Reviewer 5: nit: 

it's probably best to pin specific version (in case things move around)

```suggestion
The [Semantic Conventions for Generative AI](https://github.com/open-telemetry/semantic-conventions/tree/v1.28.0/docs/gen-ai) focus on capturing insights into AI model behavior through three primary signals: Spans, Metrics, and Events.
```

**Metrics: Monitoring Usage and Performance**  
Metrics aggregate high-level indicators like request volume, latency, and resource use, essential for managing costs and performance. This data is particularly critical for API-dependent AI applications with rate limits and cost considerations.

**Events: Capturing Detailed Interactions**  
Events log detailed moments during model execution, such as user prompts and model responses, providing a granular view of model interactions. These insights are invaluable for debugging and optimizing AI applications where unexpected behaviors may arise.

Together, these signals provide a comprehensive monitoring framework, enabling better cost management, performance tuning, and request tracing.

**Extending Observability with Vendor-Specific Attributes**

The Semantic Conventions also define vendor-specific attributes for platforms like OpenAI and Azure Inference API, ensuring telemetry captures both general and provider-specific details. This added flexibility supports multi-platform monitoring and in-depth insights. 

## Building the Python Instrumentation Library for OpenAI

> This Python-based library for OpenTelemetry captures key telemetry signals for OpenAI models, providing developers with an out-of-the-box observability solution tailored to AI workloads. The library, [hosted within the OpenTelemetry Contrib project](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation-genai/opentelemetry-instrumentation-openai-v2), automatically collects telemetry from OpenAI model interactions, including request and response metadata, token usage, and operational metrics.

Reviewer 5: another nit on pinning versions

```suggestion
This Python-based library for OpenTelemetry captures key telemetry signals for OpenAI models, providing developers with an out-of-the-box observability solution tailored to AI workloads. The library, [hosted within the OpenTelemetry Python Contrib repo](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/opentelemetry-instrumentation-openai-v2%3D%3D2.0b0/instrumentation-genai/opentelemetry-instrumentation-openai-v2), automatically collects telemetry from OpenAI model interactions, including request and response metadata and token usage.
```

no metrics yet 🤐

As generative AI applications grow, additional instrumentation libraries for other languages will follow, extending OpenTelemetry support across more tools and environments. The current library’s focus on OpenAI highlights its popularity and demand within AI development, making it a valuable initial implementation.

### Example Usage

Here’s an example of using the OpenTelemetry Python library to monitor a generative AI application with the OpenAI client.
Make sure you first install the library:

```bash
pip install opentelemetry-instrumentation-openai-v2
```

Then include the following code in your python application.

```python
from openai import OpenAI
from opentelemetry.instrumentation.openai_v2 import OpenAIInstrumentor

OpenAIInstrumentor().instrument()

> 

Reviewer 5: we probably need to add the code that configure otel SDK (or add a comment pointing to docs on how to set it up).

The code snippet:

```python
# NOTE: OpenTelemetry Python Logs and Events APIs are in beta 
from opentelemetry import trace, _logs, _events
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk._logs import LoggerProvider
from opentelemetry.sdk._events import EventLoggerProvider

from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.sdk._logs.export import BatchLogRecordProcessor
from opentelemetry.exporter.otlp.proto.grpc.trace_exporter import OTLPSpanExporter
from opentelemetry.exporter.otlp.proto.grpc._log_exporter import OTLPLogExporter

trace.set_tracer_provider(TracerProvider())
trace.get_tracer_provider().add_span_processor(
    BatchSpanProcessor(OTLPSpanExporter())
)

_logs.set_logger_provider(LoggerProvider())
_logs.get_logger_provider().add_log_record_processor(BatchLogRecordProcessor(OTLPLogExporter()))
_events.set_event_logger_provider(EventLoggerProvider())
```

an existing example that does something similar - https://github.com/open-telemetry/opentelemetry-python/blob/stable/docs/examples/logs/example.py

/cc @lzchen @xrmx in case they have some suggestions

Author: That's a lot of extra code. Should I link to that code example or is there a better doc?

Reviewer 5: let me add it to https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2988 and, once merged, let's just link it.

client = OpenAI()
response = client.chat.completions.create(
    model="gpt-4-mini",
    messages=[{"role": "user", "content": "Write a short poem on OpenTelemetry."}],
)

# The library captures telemetry, including request and response metadata, token usage, and more.
```

With this simple instrumentation, one can begin capture traces from their generative AI application. Here is an example from the Aspire Dashboard for local debugging.

![Chat trace in Aspire Dashboard](aspire_dashboard_trace.png)

> Here is a similar trace captured in Jaeger:

Reviewer 5: ```suggestion
Here is a similar trace captured in [Jaeger](https://www.jaegertracing.io/docs/next-release-v2/getting-started/#running):
```

OTEL_EXPORTER_OTLP_ENDPOINT=http://localhost:4318
OTEL_EXPORTER_OTLP_PROTOCOL=http/protobuf
OTEL_SERVICE_NAME=python-opentelemetry-openai
OPENAI_API_KEY=<replace_with_your_openai_api_key>

# Set to false or remove to disable log events

> OTEL_INSTRUMENTATION_GENAI_CAPTURE_MESSAGE_CONTENT=true

Reviewer 3: main thing here is that if you set this to false, it will crash (at least for me)

Reviewer 3: @lzchen @xrmx can one of you double-check the stack trace which happens when this is set to false or unset? I'm not sure if this is already fixed in main or not.

https://github.com/open-telemetry/opentelemetry.io/pull/5575#issuecomment-2475434466

Reviewer 3: https://github.com/open-telemetry/opentelemetry-python/pull/4276 or similar will fix this, though not sure if all backends expect a log with no body or not.
